src/main/java/org/truffleruby/core/cast/BooleanCastNode.java

```diff
@@ -80,21 +75,32 @@ protected boolean doSymbol(RubySymbol object) {
     @Specialization(guards = "isForeignObject(object)", limit = "getCacheLimit()")
     protected boolean doForeignObject(Object object,
             @CachedLibrary("object") InteropLibrary objects,
-            @Cached ConditionProfile profile,
+            @Cached ConditionProfile isNullProfile,
+            @Cached ConditionProfile isBooleanProfile,
             @Cached BranchProfile failed) {
-        if (profile.profile(objects.isBoolean(object))) {
-            try {
-                return objects.asBoolean(object);
-            } catch (UnsupportedMessageException e) {
-                failed.enter();
-                // it concurrently stopped being boolean
+
+        if (isNullProfile.profile(objects.isNull(object))) {
+            return false;
+        } else {
+            if (isBooleanProfile.profile(objects.isBoolean(object))) {
+                try {
+                    return objects.asBoolean(object);
+                } catch (UnsupportedMessageException e) {
+                    failed.enter();
+                    // it concurrently stopped being boolean
+                    return true;
+                }
+            } else {
                 return true;
             }
-        } else {
-            return true;
         }
     }
 
+    @Specialization
```

Comment:
DynamicObject specialization should be lower because a object from `isForeignObject(object)` is also `DynamicObject`:
```
ruby> r = Polyglot.eval('js', 'null')
DynamicObject<null>@50e7d8ba
```

Reply:
Order alone is not enough, because suppose for a given `if x` in some code, `x` is initially a Ruby object (then it returns true), but then on another call `x` is a foreign object which happens to be represented as a `DynamicObject`, then the node would consider it `true` too by taking the `DynamicObject` specialization (even the foreign object `isNull`). This is due to the `DynamicObject` specialization being already activated and the others not.
This was already a problem before in this node.

I can think of 2 ways to fix this:
* Group them in a single specialization. Actually no Ruby DynamicObject is isNull or isBoolean so it's fine.
* Add RubyLibrary#isForeign() and use that as a guard for both specializations.

The first one is simpler.

I might add `RubyLibrary#isForeign()` separately as it would be generally useful (`isForeignObject(object)` is quite slow, unless the class of `object` is known by the JIT).

Reply:
> Order alone is not enough, because suppose for a given if x in some code, x is initially a Ruby object (then it returns true), but then on another call x is a foreign object which happens to be represented as a DynamicObject, then the node would consider it true too by taking the DynamicObject specialization (even the foreign object isNull).

Could you give an example of how to reproduce it? I understand that `x` can be `DynamicObject` but `isForeignObject(object)` still should be `true` and "catch" the flow (as we actually have for the `Polyglot.eval('js', 'null')`), no?

Reply:
Not tested, but this should trigger it (JS's `null` is a DynamicObject):
```ruby
js_null = Polyglot.eval('js', 'null')
[Object.new, js_null].map { |x| x ? true : false } # should be [true, false], but actually [true, true]
```

Reply:
The best way to understand this is to look at the generated code (the subclass of `BooleanCastNode`) in `execute()` and `executeAndSpecialize()`.

Reply:
Done 👍